crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,506 @@
+use bevy_tasks::{ComputeTaskPool, Scope, TaskPool};
+use bevy_utils::default;
+use bevy_utils::syncunsafecell::SyncUnsafeCell;
+#[cfg(feature = "trace")]
+use bevy_utils::tracing::{info_span, Instrument};
+
+use async_channel::{Receiver, Sender};
+use fixedbitset::FixedBitSet;
+
+use crate::{
+    archetype::ArchetypeComponentId,
+    query::Access,
+    schedule_v3::{
+        is_apply_system_buffers, BoxedCondition, ExecutorKind, SystemExecutor, SystemSchedule,
+    },
+    world::World,
+};
+
+/// Per-system data used by the [`MultiThreadedExecutor`].
+// Copied here because it can't be read from the system when it's running.
+struct SystemTaskMeta {
+    /// Indices of the systems that directly depend on the system.
+    dependents: Vec<usize>,
+    /// The `ArchetypeComponentId` access of the system.
+    archetype_component_access: Access<ArchetypeComponentId>,
+    /// Is `true` if the system does not access `!Send` data.
+    is_send: bool,
+    /// Is `true` if the system is exclusive.
+    is_exclusive: bool,
+}
+
+/// Runs the schedule using a thread pool. Non-conflicting systems can run in parallel.
+pub struct MultiThreadedExecutor {
```

Comment:
Can we new-type FixedBitSet to better represent the exact invariants that we're upholding here? The logic here is both critical and tricky to understand.